File: app/src/main/java/org/carstenf/wordfinder/GameState.java
/*
 * Copyright Carsten Friedrich (Carsten.Friedrich@gmail.com)
 *
 * License: GNU GENERAL PUBLIC LICENSE 3.0 (https://www.gnu.org/copyleft/gpl.html)
 *
 */
package org.carstenf.wordfinder;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;

import android.os.CountDownTimer;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import org.jetbrains.annotations.NotNull;

public class GameState extends ViewModel {
	private Dictionary dictionary;
	private boolean autoAddPrefixalWords;

	Dictionary getDictionary() {
		return dictionary;
	}

	public void setDictionary(Dictionary dictionary) {
		this.dictionary = dictionary;
	}
	final private char[] board = new char[16];

	final WordInfoCache wordInfoCache = new WordInfoCache();

	//
	// English letter frequencies: http://en.wikipedia.org/wiki/Letter_frequency
	//
	// a 8.167%
	// b 1.492%
	// c 2.782%
	// d 4.253%
	// e 12.702%
	// f 2.228%
	// g 2.015%
	// h 6.094%
	// i 6.966%
	// j 0.153%
	// k 0.772%
	// l 4.025%
	// m 2.406%
	// n 6.749%
	// o 7.507%
	// p 1.929%
	// q 0.095%
	// r 5.987%
	// s 6.327%
	// t 9.056%
	// u 2.758%
	// v 0.978%
	// w 2.360%
	// x 0.150%
	// y 1.974%
	// z 0.074%

	// Source: https://en.wikipedia.org/wiki/Letter_frequency
	private final static double[] letterFreqProbEnglish = { 0.08167, 0.09659, 0.12441, 0.16694,
			0.29396, 0.31624, 0.33639, 0.39733, 0.46699, 0.46852, 0.47624,
			0.51649, 0.54055, 0.60804, 0.68311, 0.7024, 0.70335, 0.76322,
			0.82649, 0.91705, 0.94463, 0.95441, 0.97801, 0.97951, 0.99925, 1 };

    private final static double[] letterFreqProbGerman = { 0.068050,
            0.086910,
            0.114230,
            0.164990,
            0.339010,
            0.355570,
            0.385660,
            0.431430,
            0.496930,
            0.499610,
            0.513780,
            0.548150,
            0.573490,
            0.671250,
            0.699405,
            0.706105,
            0.706285,
            0.776315,
            0.852085,
            0.913625,
            0.960260,
            0.968720,
            0.987930,
            0.988270,
            0.988660,
            1.000000 };

	final private MutableLiveData<ArrayList<Result>> computerResultList = new MutableLiveData<>(new ArrayList<>());
	final private ArrayList<Result> playerResultList = new ArrayList<>();

	public MutableLiveData<Long> getCountDownTimerCurrentValue() {
		return countDownTimerCurrentValue;
	}

	final private MutableLiveData<Long> countDownTimerCurrentValue = new MutableLiveData<>(-1L);

	public MutableLiveData<WordInfo> getWordLookupResult() {
		return wordLookupResult;
	}

	public MutableLiveData<String> getWordLookupError() {
		return wordLookupError;
	}

	final private MutableLiveData<WordInfo> wordLookupResult = new MutableLiveData<>(null);
	final private MutableLiveData<String> wordLookupError = new MutableLiveData<>(null);

	@NonNull
	MutableLiveData<ArrayList<Result>> getComputerResultList() {
		return computerResultList;
	}

	@NonNull
    ArrayList<Result> getPlayerResultList() {
		return playerResultList;
	}

	char getBoard(int move) {
		return board[move];
	}

	final private boolean[] playerTaken = new boolean[16];

    public boolean isTimeUp() {
        return timeUp;
    }

    public void setTimeUp(boolean timeUp) {
        this.timeUp = timeUp;
    }

    private boolean timeUp = false;

	@Nullable
    private SolveTask solver;

	void shuffle() {
		clearGuess();
		setTimeUp(false);
		for (int i = 0; i < 16; i++) {
			board[i] = pickRandomLetter();
		}

		Objects.requireNonNull(computerResultList.getValue()).clear();
		computerResultList.postValue(computerResultList.getValue());
	}

	private char pickRandomLetter() {
		double r = Math.random();
		int i=0;

        double[] letterFreqProb = letterFreqProbEnglish;

		if(dictionaryName.equalsIgnoreCase("german")) {
            letterFreqProb = letterFreqProbGerman;
        }

		while(letterFreqProb[i]<r) i++;
		
		return (char) ('A'+i);
	}

	boolean findWord(String word) {
		word = word.toUpperCase().replaceAll("QU", "Q");
		boolean[] taken = new boolean[16];

        char[] chars = word.toCharArray();
		int index = 0;

		for (int i = 0; i < 16; i++)
			if (findWord(chars, index, taken, i))
				return true;
		return false;
	}

	private boolean findWord(@NonNull char[] chars, int index, boolean[] taken, int move) {
		if (taken[move])
			return false;
		if (chars[index] != board[move])
			return false;

		if (index == chars.length - 1)
			return true;

		taken[move] = true;

		for (int i = 0; i < WordFinder.MOVES[move].length; i++) {
			if (findWord(chars, index + 1, taken, WordFinder.MOVES[move][i]))
				return true;
		}

		taken[move] = false;

		return false;
	}

	void addComputerResult(Result result) {
		ArrayList<Result> list = computerResultList.getValue();
		if(list!=null) {
			list.add(result);
			Collections.sort(list, (object1, object2) -> {
				int s1 = object1.toString().length();
				int s2 = object2.toString().length();
				int res = -Double.compare(s1, s2);
				if (res == 0)
					res = object1.toString().compareTo(object2.toString());
				return res;
			});

			computerResultList.postValue(list);
		}
	}

	void stopSolving() {
		if (solver != null) {
			solver.cancel();
			solver = null;
		}
	}

	void startSolving() {
		solver = new SolveTask(this);
		solver.execute();
	}

	boolean isAvailable(int i) {
		return !playerTaken[i];
	}

	void clearGuess() {
		currentGuess = "";
		lastMove = -1;
		Arrays.fill(playerTaken, false);
	}

	@NonNull
    private String currentGuess = "";

	@NonNull
    String getCurrentGuess() {
		return currentGuess;
	}

	private int lastMove = -1;
	private boolean allow3LetterWords = true;
	@NonNull
    private SCORE_ALG scoreAlg = SCORE_ALG.COUNT;
	private String dictionaryName;

	String getDictionaryName() {
		return dictionaryName;
	}

	boolean isAllow3LetterWords() {
		return allow3LetterWords;
	}

	int getLastMove() {
		return lastMove;
	}

	void play(int move) {
		lastMove = move;
		currentGuess = currentGuess + board[move];
		playerTaken[move] = true;
	}

	public WordInfo getWordInfoFromCache(String word, String language) {
		return wordInfoCache.get(word, language);
	}

	public void processWordLookupError(String word, String language, String error) {
		wordInfoCache.put(new WordInfo(word, language, null, null));
		wordLookupError.postValue(error);
	}

	public void processWordLookupResult(WordInfo wordInfo) {
		wordInfoCache.put(wordInfo);
		wordLookupResult.postValue(wordInfo);
	}

	public boolean autoAddPrefixalWords() {
		return autoAddPrefixalWords;
	}

	public void setAutoAddPrefixalWords(boolean autoAddPrefixPref) {
		autoAddPrefixalWords = autoAddPrefixPref;
	}

	public enum PLAYER_GUESS_STATE {
		TOO_SHORT,
		ALREADY_FOUND,
		NOT_IN_DICTIONARY
	}

	PLAYER_GUESS_STATE validatePlayerGuess(@NotNull String guess) {
		int minLength = isAllow3LetterWords() ? 3 : 4;

		if(guess.length() < minLength) return PLAYER_GUESS_STATE.TOO_SHORT;

		for (Result result : playerResultList) {
			if (result.toString().equalsIgnoreCase(guess))
				return PLAYER_GUESS_STATE.ALREADY_FOUND;
		}
		if(getDictionary().lookup(guess, dictionaryName) == null) {
			return PLAYER_GUESS_STATE.NOT_IN_DICTIONARY;
		}

		return null;
	}

	void setDictionaryName(String string) {
		this.dictionaryName = string;
	}

	void setScoringAlgorithm(String string) {
		if ("count".equalsIgnoreCase(string)) {
			this.scoreAlg = SCORE_ALG.COUNT;
		} else {
			this.scoreAlg = SCORE_ALG.VALUE;
		}
	}

	void setAllow3LetterWords(boolean flag) {
		this.allow3LetterWords = flag;
		if (!isAllow3LetterWords()) {
			for (Iterator<Result> iter = playerResultList.iterator(); iter
					.hasNext();) {
				Result next = iter.next();
				if (next.toString().length() < 4)
					iter.remove();
			}

			ArrayList<Result> crl = computerResultList.getValue();
			if(crl!=null) {
				for (Iterator<Result> iter = crl.iterator(); iter
						.hasNext(); ) {
					Result next = iter.next();
					if (next.toString().length() < 4)
						iter.remove();
				}
				computerResultList.postValue(crl);
			}
		}
	}

	@Nullable
    private CountDownTimer countDownTimer = null;

    public long getCountDownTime() {
        return countDownTime;
    }

    private long countDownTime = -1;

	void setCountDownTime(long time) {
		countDownTime = time;
	}

	void startCountDown() {
		if (countDownTimer != null)
            countDownTimer.cancel();

		if (countDownTime < 0)
			return;

		countDownTimerCurrentValue.postValue(countDownTime);

		countDownTimer = new CountDownTimer(countDownTime, 1000) {

			public void onTick(long millisUntilFinished) {
				countDownTimerCurrentValue.postValue(millisUntilFinished / 1000);
			}

			public void onFinish() {
				setTimeUp(true);
				countDownTimerCurrentValue.postValue(-1L);
			}
		}.start();

	}

	private int getScore(List<Result> list) {
		if(list==null) return 0;

		int res = 0;
		switch (this.scoreAlg) {
		case COUNT:
			res = list.size();
			break;
		case VALUE:
			for (Result result : list) {
				switch (result.toString().length()) {
				case 3:
				case 4:
					res += 1;
					break;
				case 5:
					res += 2;
					break;
				case 6:
					res += 3;
					break;
				case 7:
					res += 5;
					break;
				default:
					res += 11;
				}
			}
			break;
		}
		return res;
	}

	int getPlayerScore() {
		return getScore(playerResultList);
	}

    public boolean hasGameStarted() {
	    return board[0] != 0;
    }

    enum SCORE_ALG {
		COUNT, VALUE
	}

	int getComputerScore() {
		return getScore(computerResultList.getValue());
	}
}

File: app/src/main/kotlin/org/carstenf/wordfinder/GameState.kt
/*
 * Copyright Carsten Friedrich (Carsten.Friedrich@gmail.com)
 *
 * License: GNU GENERAL PUBLIC LICENSE 3.0 (https://www.gnu.org/copyleft/gpl.html)
 *
 */
package org.carstenf.wordfinder

import android.os.CountDownTimer
import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import java.util.Arrays
import java.util.Locale

class GameState : ViewModel() {
    var solveFinished: Boolean = false
    private var letterSelector: LETTER_RANDOM_DIST? = LETTER_RANDOM_DIST.MULTI_LETTER_FREQUENCY

    var timerMode: TIMER_MODE = TIMER_MODE.COUNT_DOWN

    lateinit var dictionary: Dictionary

    private var autoAddPrefixalWords = false

    private val board = CharArray(16)

    val computerResultList: MutableLiveData<ArrayList<Result>?> = MutableLiveData(ArrayList())
    val playerResultList: ArrayList<Result> = ArrayList()

    val timerCurrentValue: MutableLiveData<Long> = MutableLiveData(0L)

    fun getBoard(move: Int): Char {
        return board[move]
    }

    private val playerTaken = BooleanArray(16)

    private var solver: SolveTask? = null

    fun shuffle() {
        clearGuess()
        gameLifecycleState.postValue(GameLifeCycleState.STARTED)

        val letterCounts = IntArray(26)

        var letterRandomDistCur = letterSelector
        if(letterRandomDistCur == null) {
            Log.d(WordFinder.TAG, "Letter selector: Whatever")
            letterRandomDistCur =  LETTER_RANDOM_DIST.entries.toTypedArray().random()
        }

        Log.d(WordFinder.TAG, "Letter selector: ${letterRandomDistCur.name}")

        for (i in 0..15) {
            board[i] = pickRandomLetter(letterRandomDistCur, letterCounts, dictionaryCountryCode(), i).uppercaseChar()
        }

        board.shuffle()

        computerResultList.value?.clear()
        computerResultList.postValue(computerResultList.value)
    }

    fun findWord(w: String): Boolean {
        var word = w
        word = word.uppercase(Locale.getDefault()).replace("QU".toRegex(), "Q")
        val taken = BooleanArray(16)

        val chars = word.toCharArray()
        val index = 0

        for (i in 0..15) if (findWord(chars, index, taken, i)) return true
        return false
    }

    private fun findWord(chars: CharArray, index: Int, taken: BooleanArray, move: Int): Boolean {
        if (taken[move]) return false
        if (chars[index] != board[move]) return false

        if (index == chars.size - 1) return true

        taken[move] = true

        for (i in WordFinder.MOVES[move].indices) {
            if (findWord(chars, index + 1, taken, WordFinder.MOVES[move][i])) return true
        }

        taken[move] = false

        return false
    }

    fun addComputerResult(result: Result) {
        val list = computerResultList.value
        if (list != null) {
            list.add(result)
            list.sortWith(Comparator { object1: Result, object2: Result ->
                val s1 = object1.toString().length
                val s2 = object2.toString().length
                var res = -s1.toDouble().compareTo(s2.toDouble())
                if (res == 0) res = object1.toString().compareTo(object2.toString())
                res
            })

            computerResultList.postValue(list)
        }
    }

    fun stopSolving() {
        solver?.cancel()
        solveFinished = true
        solver = null
    }

    fun startSolving() {
        solver = SolveTask(this)
        solveFinished = false
        solver?.execute()
    }

    fun isAvailable(i: Int): Boolean {
        return !playerTaken[i]
    }

    fun clearGuess() {
        currentGuess = ""
        lastMove = -1
        Arrays.fill(playerTaken, false)
    }

    var currentGuess: String = ""
        private set

    var lastMove: Int = -1
        private set

    var isAllow3LetterWords: Boolean = true
        set(flag) {
            field = flag
            if (!isAllow3LetterWords) {
                val iter =
                    playerResultList.iterator()
                while (iter
                        .hasNext()
                ) {
                    val next = iter.next()
                    if (next.toString().length < 4) iter.remove()
                }

                val crl =
                    computerResultList.value
                if (crl != null) {
                    val iter2 =
                        crl.iterator()
                    while (iter2
                            .hasNext()
                    ) {
                        val next = iter2.next()
                        if (next.toString().length < 4) iter2.remove()
                    }
                    computerResultList.postValue(crl)
                }
            }
        }

    private var scoreAlg = ScoreAlgorithm.COUNT

    var dictionaryName: String? = null

    fun play(move: Int) {
        lastMove = move
        currentGuess += board[move]
        playerTaken[move] = true
    }

    fun autoAddPrefixalWords(): Boolean {
        return autoAddPrefixalWords
    }

    fun setAutoAddPrefixalWords(autoAddPrefixPref: Boolean) {
        autoAddPrefixalWords = autoAddPrefixPref
    }

    enum class PlayerGuessState {
        TOO_SHORT,
        ALREADY_FOUND,
        NOT_IN_DICTIONARY,
        GUESS_VALID
    }

    val gameLifecycleState: MutableLiveData<GameLifeCycleState> =
        MutableLiveData(GameLifeCycleState.NOT_STARTED)

    enum class GameLifeCycleState {
        NOT_STARTED,
        STARTED,
        TIMER_STARTED,
        UNSOLVABLE,
        TIMER_FINISHED,
        GAME_OVER
    }

    enum class LETTER_RANDOM_DIST {
        UNIFORM,
        LETTER_FREQUENCY,
        MULTI_LETTER_FREQUENCY,
        LETTER_DICE
    }

    enum class TIMER_MODE {
        COUNT_DOWN,
        STOP_WATCH
    }

    data class PlayerGuessResult(val guess: Dictionary.WordInfoData, val state: PlayerGuessState)

    suspend fun validatePlayerGuess(guess: String): PlayerGuessResult {
        val minLength = if (isAllow3LetterWords) 3 else 4

        if (guess.length < minLength) return PlayerGuessResult(
            Dictionary.WordInfoData(guess),
            PlayerGuessState.TOO_SHORT
        )

        val lookupResult = dictionary.lookup(guess, dictionaryName)
            ?: return PlayerGuessResult(
                Dictionary.WordInfoData(guess),
                PlayerGuessState.NOT_IN_DICTIONARY
            )

        for (result in playerResultList) {
            if (result.toString()
                    .equals(lookupResult.displayText, ignoreCase = true)
            ) return PlayerGuessResult(
                Dictionary.WordInfoData(guess),
                PlayerGuessState.ALREADY_FOUND
            )
        }

        return PlayerGuessResult(lookupResult, PlayerGuessState.GUESS_VALID)
    }

    fun setScoringAlgorithm(string: String?) {
        if ("count".equals(string, ignoreCase = true)) {
            this.scoreAlg = ScoreAlgorithm.COUNT
        } else {
            this.scoreAlg = ScoreAlgorithm.VALUE
        }
    }

    private var countDownTimer: CountDownTimer? = null
    var countUpTimer: CountUpTimer? = null

    var gameTime: Long = -1

    fun startTimer() {
        startTimer(gameTime)
    }

    private fun startTimer(time: Long) {
        cancelTimer()

        if (gameTime < 0) return

        gameLifecycleState.postValue(GameLifeCycleState.TIMER_STARTED)
        timerCurrentValue.postValue(time / 1000)

        if(timerMode == TIMER_MODE.COUNT_DOWN) {
            countDownTimer = object : CountDownTimer(time, 1000) {
                override fun onTick(millisUntilFinished: Long) {
                    timerCurrentValue.postValue(millisUntilFinished / 1000)
                }

                override fun onFinish() {
                }
            }.start()
        } else {
            countUpTimer =  CountUpTimer{ seconds: Long -> timerCurrentValue.postValue(seconds)}
            countUpTimer?.start()
        }

    }

    private fun getScore(list: List<Result>?): Int {
        if (list == null) return 0

        var res = 0
        when (this.scoreAlg) {
            ScoreAlgorithm.COUNT -> res = list.size
            ScoreAlgorithm.VALUE -> for (result in list) {
                res += when (result.toString().length) {
                    3, 4 -> 1
                    5 -> 2
                    6 -> 3
                    7 -> 5
                    else -> 11
                }
            }
        }
        return res
    }

    fun onPause() {
        // If game is interrupted while in progress, cancel timer and restart later in on Resume()
        if (gameLifecycleState.value == GameLifeCycleState.TIMER_STARTED) {
            countDownTimer?.cancel()
            countUpTimer?.cancel()
        }
    }

    fun onResume() {
        val state = gameLifecycleState.value ?: return

        // If game was interrupted while in progress continue with the left over time.
        if (state == GameLifeCycleState.TIMER_STARTED) {
            timerCurrentValue.value?.let { startTimer(it * 1000L) }
        } else if (state >= GameLifeCycleState.TIMER_FINISHED) {
            gameLifecycleState.postValue(state) // Give UI a chance to update as well.
        }
    }

    fun cancelTimer() {
        countDownTimer?.cancel()
        countUpTimer?.cancel()
    }

    fun onSolveFinished() {
        if (computerResultList.value?.size == 0) {
            cancelTimer()
            gameLifecycleState.postValue(GameLifeCycleState.UNSOLVABLE)
        }
        solveFinished = true
    }

    fun dictionaryCountryCode(): LANGUAGE {
        if (dictionaryName.equals("german", ignoreCase = true) ||
            dictionaryName.equals("german_wiki", ignoreCase = true) ||
            dictionaryName.equals("german_simple", ignoreCase = true)) {
            return LANGUAGE.DE
        }
        return LANGUAGE.EN
    }

    fun setLetterSelector(distStr: String?) {
        if ("uniformRandom".equals(distStr, ignoreCase = true)) {
            letterSelector = LETTER_RANDOM_DIST.UNIFORM
        } else if ("letterFrequency".equals(distStr, ignoreCase = true)) {
            letterSelector = LETTER_RANDOM_DIST.LETTER_FREQUENCY
        } else if ("multiLetterFrequence".equals(distStr, ignoreCase = true)) {
            letterSelector = LETTER_RANDOM_DIST.MULTI_LETTER_FREQUENCY
        } else if ("letterDice".equals(distStr, ignoreCase = true)) {
            letterSelector = LETTER_RANDOM_DIST.LETTER_DICE
        }  else if ("whateverRandom".equals(distStr, ignoreCase = true)) {
            letterSelector = null
        }
    }

    val playerScore: Int
        get() = getScore(playerResultList)

    internal enum class ScoreAlgorithm {
        COUNT, VALUE
    }

    val computerScore: Int
        get() = getScore(computerResultList.value)

}